src/hclust.jl

```diff
@@ -0,0 +1,452 @@
+## hclust.jl (c) 2014 David A. van Leeuwen 
+## Hierarchical clustering, similar to R's hclust()
+
+## Algorithms are based upon C. F. Olson, Parallel Computing 21 (1995) 1313--1325. 
+
+## This seems to work like R's implementation, but it is extremely inefficient
+## This probably scales O(n^3) or worse. We can use it to check correctness
+
+## This is also in types.jl, but that is not read...
+## Mostly following R's hclust class
+type Hclust{T<:Real}
+    merge::Matrix{Int}
+    height::Vector{T}
+    order::Vector{Int}
+    labels::Vector
+    method::Symbol
+end
+
+function hclust_n3{T<:Real}(d::Matrix{T}, method::Function)
+    @assert size(d,1) == size(d,2)
+    if !issym(d)
+        d += d'                 # replaces d, which must be symmetric
+    end
+    mr = Int[]                  # min row
+    mc = Int[]                  # min col
+    h = T[]                     # height
+    nc = size(d,1)              # number of clusters
+    cl = -[1:nc]                # segment to cluster attribution, initially negative
+    next = 1                    # next cluster label
+    while next < nc 
+        min = Inf
+        mi = mj = 0
+        cli = unique(cl)
+        mask = BitVector(nc)
+        for j in 1:length(cli)           # loop over for lower triangular indices, i>j
+            cols = cl .== cli[j]
+            for i in (j+1):length(cli)
+                rows = cl.==cli[i]
+                distance = method(d[rows,cols])
+                if distance < min
+                    min = distance
+                    mi = cli[i]
+                    mj = cli[j]
+                    mask = cols | rows
+                end
+            end
+        end 
+        ## simulate R's order
+        if mi < 0 && mj < 0 && mi > mj ||
+            mi > 0 && mj > 0 && mi < mj ||
+            mi < 0 && mj > 0
+            push!(mr, mi)
+            push!(mc, mj)
+        else
+            push!(mr, mj)
+            push!(mc, mi)
+        end
+        push!(h, min)
+        cl[mask] = next
+        next += 1
+    end
+    hcat(mr, mc), h
+end
+
+## Efficient single link algorithm, according to Olson, O(n^2), fig 2. 
+## Verified against R's implementation, correct, and about 2.5 x faster
+## For each i < j compute D(i,j) (this is already given)
+## For each 0 < i <= n compute Nearest Neighbor N(i)
+## Repeat n-1 times
+##   find i,j that minimize D(i,j)
+##   merge clusters i and j
+##   update D(i,j) and N(i) accordingly
+function hclust_minimum{T<:Real}(d::Matrix{T})
+    @assert size(d,1) == size(d,2)
+    ## For each i < j compute d[i,j] (this is already given) --- we need a local copy
+    if !issym(d)
```

Comment:
We could just take a lower trianglular matrix. It would be the users
responsibility to ensure that the distance function they computed was
symmetric. Taking the triangular input also saves some time for the user.

On Thu, Mar 19, 2015, 4:26 AM David van Leeuwen <notifications@github.com>
wrote:

> In src/hclust.jl
> <https://github.com/JuliaStats/Clustering.jl/pull/32#discussion_r26735776>
> :
>
> > +    end
> > +    hcat(mr, mc), h
> > +end
> > +
> > +## Efficient single link algorithm, according to Olson, O(n^2), fig 2.
> > +## Verified against R's implementation, correct, and about 2.5 x faster
> > +## For each i < j compute D(i,j) (this is already given)
> > +## For each 0 < i <= n compute Nearest Neighbor N(i)
> > +## Repeat n-1 times
> > +##   find i,j that minimize D(i,j)
> > +##   merge clusters i and j
> > +##   update D(i,j) and N(i) accordingly
> > +function hclust_minimum{T<:Real}(d::Matrix{T})
> > +    @assert size(d,1) == size(d,2)
> > +    ## For each i < j compute d[i,j] (this is already given) --- we need a local copy
> > +    if !issym(d)
>
> I am not too happy with issym() myself as it takes some time.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/JuliaStats/Clustering.jl/pull/32/files#r26735776>.
>